### PR TITLE
extensions: Graphite stats sink

### LINKED ISSUE #16523: Support sending stats with Graphite-formatted tags
*Title*: *Support sending stats with Graphite-formatted tags*

*Description*:
I'm currently working on a Graphite tag-aware stat sink, and we'd love to get this extension accepted into the upstream Envoy repo. This is a follow-up to issue #14863 but smaller in scope (#14863 originally proposed a more flexible sink, while what I'm proposing here would be limited to Graphite tags).

*Relevant Links*:
https://graphite.readthedocs.io/en/latest/tags.html describes Graphite tags.

### PR DESCRIPTION
**Pull request by vaccarium**
Adds an extension that sends Statsd metrics with Graphite-formatted tags over UDP.

Signed-off-by: Pavel Fomin <pafomin@avito.ru>

Risk Level: Low (adding a small optional feature)
Testing: Unit test and manual testing
Docs Changes: Included inline proto docs
Release Notes: Included (but probably need reviewing, not sure I got the link to docs right) 
Platform Specific Features: N/A

Fixes #16523

### CODE REVIEW

**Comment by vaccarium** [comment]
envoy-presubmit (precheck format) has failed with this output:

```
ERROR: New directory extensions/stat_sinks/graphite_statsd appears to not have owners in CODEOWNERS
```

I'm not sure how I can address this.

**Comment by lizan** [comment]
You need an entry in CODEOWNERS file, like https://github.com/envoyproxy/envoy/blob/main/CODEOWNERS#L126

File: api/envoy/config/metrics/v3/stats.proto
Comment (lizan) [inline api/envoy/config/metrics/v3/stats.proto:416]: No need to `reserved 2`, this exist in other proto from historical reason.

File: api/envoy/config/metrics/v3/stats.proto
Comment (lizan) [inline api/envoy/config/metrics/v3/stats.proto:418]: Why this need to be oneof, will there be more specifier later?

**Review by lizan — COMMENTED** [review]
API review

File: source/extensions/stat_sinks/graphite_statsd/BUILD
Comment (phlax) [inline source/extensions/stat_sinks/graphite_statsd/BUILD:19]: hi @vaccarium - you will need to add an entry in `source/extensions/extensions_metadata.yaml` for this extension and re/move `category`, `security_posture` and `status` to there

**Comment by vaccarium** [comment]
> You need an entry in CODEOWNERS file, like https://github.com/envoyproxy/envoy/blob/main/CODEOWNERS#L126

Thank you! I'm not sure whom to put in as code owners, maybe someone could help? I would have copied the owners of extensions/stats_sinks/common/statsd but there are none.

File: api/envoy/config/metrics/v3/stats.proto
Comment (vaccarium) [reply api/envoy/config/metrics/v3/stats.proto:418]: Yes, TCP might become supported too in the future.

File: source/extensions/stat_sinks/graphite_statsd/BUILD
Comment (vaccarium) [reply source/extensions/stat_sinks/graphite_statsd/BUILD:19]: Thank you!

**Comment by vaccarium** [comment]
(sorry for the force-push, I've forgotten to sign off on the second commit)

File: docs/root/version_history/current.rst
Comment (phlax) [inline docs/root/version_history/current.rst:82]: hi @vaccarium - this wont work - its linking to the published docs - which dont include this - it should work with no prefix - ie linking to the current docs build - if its not it is more likely that the link itself is wrong

File: api/envoy/config/metrics/v3/stats.proto
Comment (phlax) [inline api/envoy/config/metrics/v3/stats.proto:410]: could you make this ` ``envoy.stat_sinks.graphite_statsd`` ` - using `*` is wrong it should be a literal

File: docs/root/version_history/current.rst
Comment (phlax) [reply docs/root/version_history/current.rst:82]: ..altho it looks correct to me

Comment (phlax) [reply docs/root/version_history/current.rst:82]: it seems like the previous failure was due to `statssink` rather than `statsdsink`

` `Graphite-formatted tag <envoy_v3_api_msg_config.metrics.v3.GraphiteStatsdSink>` ` should work

Comment (vaccarium) [reply docs/root/version_history/current.rst]: Thanks, I've removed the version prefix

File: api/envoy/config/metrics/v3/stats.proto
Comment (vaccarium) [reply api/envoy/config/metrics/v3/stats.proto]: done

**Comment by vaccarium** [comment]
@lizan @mattklein123 I think this PR is now ready for review.

As for the failing coverage check, it's looking like a coverage bug. The [report](https://storage.googleapis.com/envoy-pr/c552fdd/coverage/source/extensions/stat_sinks/graphite_statsd/config.cc.gcov.html) shows that the only lines not tested are unreachable. Not sure why they are not ignored though, because the unreachable branch is marked with `NOT_REACHED_GCOVR_EXCL_LINE`.

Besides, I'm still not sure who should be the code owners for the new directory, so I've added it to `UNOWNED_EXTENSIONS` for now.

**Comment by PAFomin-at-avito** [comment]
Mac CI passed, the only remaining failing check is the coverage/

**Comment by vaccarium** [comment]
Mac CI passed, the only failing check that remains is the coverage.

**Comment by phlax** [comment]
@vaccarium that seems to be a genuine CI fail - ie not a flake

> Code coverage for source/extensions/stat_sinks/graphite_statsd is lower than limit of 96.6 (85.2)

https://dev.azure.com/cncf/envoy/_build/results?buildId=77277&view=logs&j=5035ce29-be3c-5fcf-dc50-abb6a265f8b1&t=548c3e59-0417-5971-4f1b-f8545005c291&l=18038

**Comment by vaccarium** [comment]
> @vaccarium that seems to be a genuine CI fail - ie not a flake
> 
> > Code coverage for source/extensions/stat_sinks/graphite_statsd is lower than limit of 96.6 (85.2)
> 
> https://dev.azure.com/cncf/envoy/_build/results?buildId=77277&view=logs&j=5035ce29-be3c-5fcf-dc50-abb6a265f8b1&t=548c3e59-0417-5971-4f1b-f8545005c291&l=18038

Looking at the report (https://storage.googleapis.com/envoy-pr/c552fdd/coverage/source/extensions/stat_sinks/graphite_statsd/config.cc.gcov.html), the only lines not covered are those actually unreachable. The check only failed because the file in question is very short, with only 27 significant lines of code.

**Comment by vaccarium** [comment]
(BTW, I expected that a line containing `NOT_REACHED_GCOVR_EXCL_LINE` would be ignored by the coverage checker, but apparently it's not the case)

**Comment by phlax** [comment]
> The check only failed because the file in question is very short

ah, k - i see

> I expected that a line containing NOT_REACHED_GCOVR_EXCL_LINE would be ignored by the coverage checker, but apparently it's not the case

hmm - yep - i was thinking the same

beyond my immediate knowledge im afraid

**Comment by mattklein123** [comment]
Regarding ownership, @vaccarium you will need to be an owner. If you are not willing to do that we can't merge this. You can put me as the other owner, but if you are not response to bugs, reviews, etc. we will remove the extension.

Regarding coverage, if the only lines not covered are the not reached lines, just adjust the coverage minimum accordingly.

/wait

File: source/extensions/stat_sinks/common/statsd/statsd.cc
Comment (mattklein123) [inline source/extensions/stat_sinks/common/statsd/statsd.cc:87]: This logic is duplicated in 3 places, please collapse this into a utility somehow.

**Comment by vaccarium** [comment]
Extremely sorry to everyone auto-added as reviewers, I've messed up a DCO fix :(

**Comment by mattklein123** [comment]
Please check CI and then I can take a look.

/wait

**Comment by vaccarium** [comment]
Hmm, I still cannot get one Windows test happy:
```
FAILED: //test/extensions/resource_monitors/injected_resource:injected_resource_monitor_integration_test (Summary)
      C:/build/tmp/execroot/envoy/bazel-out/x64_windows-opt/testlogs/test/extensions/resource_monitors/injected_resource/injected_resource_monitor_integration_test/test.log
      C:/build/tmp/execroot/envoy/bazel-out/x64_windows-opt/testlogs/test/extensions/resource_monitors/injected_resource/injected_resource_monitor_integration_test/test_attempts/attempt_1.log
INFO: From Testing //test/extensions/resource_monitors/injected_resource:injected_resource_monitor_integration_test:
==================== Test output for //test/extensions/resource_monitors/injected_resource:injected_resource_monitor_integration_test:
[symbolize_win32.inc : 53] RAW: SymInitialize() failed: 2147483653
```
From looking at the post-submit pipeline ([example](https://dev.azure.com/cncf/envoy/_build/results?buildId=77759&view=logs&j=b840a642-5ff3-5357-2e4b-e06e40b0cffd&t=67965174-5100-5631-9dc0-68b9f0aacb53)) it seems like a common problem right now. @wrowe sorry to bother you but maybe you could give some pointers on this?

**Comment by mattklein123** [comment]
Windows CI is broken. You can ignore it for now.

File: api/envoy/config/metrics/v3/stats.proto
Comment (mattklein123) [inline api/envoy/config/metrics/v3/stats.proto:415]: For historical reasons, all of the stat sinks are here. However for this new extension, can you add it where it probably should be, which is in this tree: https://github.com/envoyproxy/envoy/tree/main/api/envoy/extensions

I would just make a dedicated file for it.

File: source/extensions/stat_sinks/common/statsd/statsd.cc
Comment (mattklein123) [inline source/extensions/stat_sinks/common/statsd/statsd.cc:148]: You shouldn't need the default as every branch is covered. If this is still failing to compile, move the `NOT_REACHED_GCOVR_EXCL_LINE` to the end of the function.

File: source/extensions/stat_sinks/common/statsd/tag_formats.h
Comment (mattklein123) [inline source/extensions/stat_sinks/common/statsd/tag_formats.h:16]: I would either make these full std::string, or constexpr absl::string_view.

File: source/extensions/stat_sinks/common/statsd/tag_formats.h
Comment (mattklein123) [inline source/extensions/stat_sinks/common/statsd/tag_formats.h:20]: Yes please, same below.

File: source/extensions/stat_sinks/graphite_statsd/config.cc
Comment (mattklein123) [inline source/extensions/stat_sinks/graphite_statsd/config.cc:13]: These files are getting removed so I would just inline the name for this.

File: test/per_file_coverage.sh
Comment (mattklein123) [inline test/per_file_coverage.sh:52]: Are you missing real coverage here? Or just the gcovr excluded issues?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks, LGTM with small comments.

/wait

File: source/extensions/stat_sinks/common/statsd/tag_formats.h
Comment (vaccarium) [reply source/extensions/stat_sinks/common/statsd/tag_formats.h:20]: Tried to use `CONSTRUCT_ON_FIRST_USE` here but for some reason the resulting code couldn't compile. Interestingly, expanding the macro manually did help. I've saved the error message and code [as a gist](https://gist.github.com/vaccarium/6409c255efaeabc23e5f547f7bcad19a), could someone more familiar with the codebase please take a look?

File: docs/root/configuration/other_features/graphite_statsd_stat_sink.rst
Comment (phlax) [inline docs/root/configuration/other_features/graphite_statsd_stat_sink.rst:4]: underline should match length of title in rst

Comment (vaccarium) [reply docs/root/configuration/other_features/graphite_statsd_stat_sink.rst:4]: thank you!

**Comment by phlax** [comment]
@vaccarium docs are generated for all api configs - which have to be linked into a `toc` somewhere - afaict the remaining docs build issue is because the api/proto isnt linked in

im not 100% certain about how this should be done - but i think probs take a look at `docs/root/api-v3/config/wasm/wasm.rst`

you probably want to do something similar - ie create a `toc` in `docs/root/api-v3/config/graphite/graphite.rst`

cc @htuch

**Comment by vaccarium** [comment]
@phlax thank you, will try this now

File: source/extensions/stat_sinks/common/statsd/statsd.cc
Comment (vaccarium) [reply source/extensions/stat_sinks/common/statsd/statsd.cc]: Done. Do you think I should remove the `default` [here](https://github.com/envoyproxy/envoy/pull/16703/files#diff-a50f6997e1b3871b731aeb52e776bf91449bdc5996227ccbc89f3e62b91827b6R39) as well?

File: test/per_file_coverage.sh
Comment (vaccarium) [reply test/per_file_coverage.sh:52]: It's mostly unreachable branches and closing braces. There's also one method in source/extensions/stat_sinks/common/statsd/statsd.cc that has no coverage in the main branch, I haven't touched it in this PR.

Relevant coverage reports (they're a couple of commits old through):
[source/extensions/stat_sinks/common/statsd/statsd.cc](https://storage.googleapis.com/envoy-pr/b66eb3b/coverage/source/extensions/stat_sinks/common/statsd/statsd.cc.gcov.html)
[source/extensions/stat_sinks/graphite_statsd/config.cc](https://storage.googleapis.com/envoy-pr/b66eb3b/coverage/source/extensions/stat_sinks/graphite_statsd/config.cc.gcov.html)

**Comment by vaccarium** [comment]
tsan seems to be flaking, sadly I cannot reproduce the issue on my machine.

File: docs/root/configuration/other_features/graphite_statsd_stat_sink.rst
Comment (mattklein123) [inline docs/root/configuration/other_features/graphite_statsd_stat_sink.rst:1]: doc nit: I don't think this is an "other feature" can you move this to the observability section in whatever way makes the most sense?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small doc comment.

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**
